codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/DocumentationConverter.java

```diff
@@ -119,33 +134,102 @@ public void head(Node node, int depth) {
             }
         }
 
-        private void writeNewline() {
-            // While jsoup will strip out redundant whitespace, it will still leave some. If we
-            // start a new line then we want to make sure we don't keep any prefixing whitespace.
-            shouldStripPrefixWhitespace = true;
-            writer.write("");
-        }
-
         private void writeText(TextNode node) {
             if (node.isBlank()) {
                 return;
             }
 
             // Docs can have valid $ characters that shouldn't run through formatters.
             String text = node.text().replace("$", "$$");
-
             if (shouldStripPrefixWhitespace) {
                 shouldStripPrefixWhitespace = false;
                 text = StringUtils.stripStart(text, " \t");
             }
 
-            if (needsListPrefix) {
-                needsListPrefix = false;
-                writer.write("");
-                writeIndent();
-                text = "* " + StringUtils.stripStart(text, " \t");
+            if (listDepth > 0) {
+                if (needsListPrefix) {
+                    needsListPrefix = false;
+                    text = "    - " + StringUtils.stripStart(text, " \t");
+                    writeNewline();
+                } else {
+                    text = StringUtils.stripStart(text, " \t");
+                }
+
+                writeText(text, "\n    ");
+            } else {
+                writeText(text, "\n");
+            }
+        }
+
+        private void writeText(String text, String newLineIndent) {
+            // check the last line's remaining space to see if test should be
+            // written to current line or new line
+            // note that wrapped text will not contain desired indent at the beginning,
+            // so indent will be added to the wrapped text if it is written to a new line
+
+            // if the last line text has reached docWrapLength, directly write at new line
+            if (lastLineRemaining <= 0) {
```

Comment:
So this is the main component of this PR and I'm confident we can simplify it a fair bit.

Put into words, the algorithm is basically "write until the current line is full, then wrap and write the rest". That statement is basically expressed directly in code in your third "branch" of this method.

That said, can't we effectively eliminate these first two branches if we fold the considerations they make into the third?

The pseudocode I'm thinking of is something like
1. partition the text into two strings A and B: text that fits on the current line (using your memory of the last written text to do so), and the remainder
2. if there's any text in A, write it and a newline
3. if there's any text in B, write it

You might also consider automatically writing a newline when you write text B and detect that it reached the wrap limit, that _may_ simplify your partitioning code somewhat (it basically totally eliminates your first branch), but I can't be certain without the code in front of me.